pkg/object/mqttproxy/broker.go

```diff
@@ -0,0 +1,336 @@
+		return
+	}
+	connect, ok := packet.(*packets.ConnectPacket)
+	if !ok {
+		logger.Errorf("mqtt.handleConn: received %s that was not Connect", packet.String())
+		return
+	}
+
+	connack := packets.NewControlPacket(packets.Connack).(*packets.ConnackPacket)
+	connack.SessionPresent = connect.CleanSession
+	connack.ReturnCode = connect.Validate()
+	if connack.ReturnCode != packets.Accepted {
+		err = connack.Write(conn)
+		logger.Errorf("mqtt.handleConn: invalid connection %v, write connack failed, err:%s", connack.ReturnCode, err)
+		return
+	}
+
+	if !b.checkClientAuth(connect) {
+		connack.ReturnCode = packets.ErrRefusedNotAuthorised
+		err = connack.Write(conn)
+		logger.Errorf("mqtt.handleConn: invalid connection %v, connack back failed, err:%s", connack.ReturnCode, err)
+		return
+	}
+
+	err = connack.Write(conn)
+	if err != nil {
+		logger.Errorf("mqtt.handleConn: send connack to client %s failed, err %s", connect.ClientIdentifier, err)
+		return
+	}
+
+	client := newClient(connect, b, conn)
+	cid := client.info.cid
+
+	b.Lock()
+	if oldClient, ok := b.clients[cid]; ok {
+		go oldClient.close()
+	}
+	b.clients[client.info.cid] = client
+	b.setSession(client, connect)
+	b.Unlock()
+
+	client.session.updateEGName(b.egName, b.name)
+	topics, qoss, _ := client.session.allSubscribes()
+	if len(topics) > 0 {
+		err = b.topicMgr.subscribe(topics, qoss, client.info.cid)
+		if err != nil {
+			logger.Errorf("client <%v> use previous session topics <%v> to subscribe failed, err:%v", client.info.cid, topics, err)
+		}
+	}
+	client.readLoop()
+}
+
+func (b *Broker) setSession(client *Client, connect *packets.ConnectPacket) {
+	// when clean session is false, previous session exist and previous session not clean session,
+	// then we use previous session, otherwise use new session
+	prevSess := b.sessMgr.get(connect.ClientIdentifier)
+	if !connect.CleanSession && (prevSess != nil) && !prevSess.cleanSession() {
+		// prevSess.update(connect)
+		client.session = prevSess
+	} else {
+		if prevSess != nil {
+			prevSess.close()
+		}
+		client.session = b.sessMgr.newSessionFromConn(connect)
+	}
+}
+
+func (b *Broker) requestTransfer(egName, name string, data HTTPJsonData) {
+	urls, err := b.memberURL(egName, name)
+	if err != nil {
+		logger.Errorf("mqtt.requestTransfer: not find url for eg:%s, name:%s, err:%v", egName, name, err)
+		return
+	}
+	jsonData, err := json.Marshal(data)
+	if err != nil {
+		logger.Errorf("mqtt.requestTransfer: json data marshal failed, err: %v", err)
+		return
+	}
+	for _, url := range urls {
+		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+		if err != nil {
+			logger.Errorf("mqtt.requestTransfer: make new request failed, err:%v", err)
+			return
+		}
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			logger.Errorf("mqtt.requestTransfer: http client send msg failed, err:%v", err)
+			return
+		}
+		defer resp.Body.Close()
```

Comment:
`defer` inside a loop is not a good practice.

Reply:
Oops, sorry for that. Shame on me.